[3.2] Move basic types to the datasource component

Marlinc opened the pull request.

Marlinc commented:
This is still work in progress but very open to discussion :smile:

lorenzo commented:
The basic types implementation is goinggn to be deprecated soon, so I would not touch the code around it too much for now

Marlinc commented:
@lorenzo yea I know that. I'm waiting for your PR to go in :+1:

Marlinc commented:
Its not ready for comments since #7564 has been merged :+1:

Marlinc commented:
Its now ready for comments since #7564 has been merged :+1:

markstory commented on src/Database/Type/BinaryType.php:
```diff
@@ -51,7 +39,7 @@ public function toDatabase($value, Driver $driver)
      * @throws \Cake\Core\Exception\Exception
      */
-    public function toPHP($value, Driver $driver)
+    public function fromDatabasetoPHP($value, Driver $driver)
```
Not sure why this method needed to be renamed. Its also making the Datasource and Database packages have different methods if I'm understanding things correctly.

markstory commented on src/Database/TypeTrait.php:
```diff
@@ -0,0 +1,159 @@
+     * @param Driver $driver object from which database preferences and configuration will be extracted
+     * @return mixed
+     */
+    public function toDatabase($value, Driver $driver)
```
I don't see why the `toDatabase` method couldn't be a simple alias for `toDatasource`?  I don't think its strange for the database API to use a method with 'datasource' in it.

markstory commented on src/Datasource/Type.php:
```diff
@@ -0,0 +1,146 @@
+        'string' => ['callback' => ['\Cake\Datasource\Type', 'strval']],
+        'text' => ['callback' => ['\Cake\Datasource\Type', 'strval']],
+        'boolean' => [
+            'callback' => ['\Cake\Datasource\Type', 'boolval']
```
Why not use the BoolType and StringType classes @lorenzo added?

markstory replied:
Couldn't the existing catalog of types be useful in other datasources as well?

Marlinc replied:
I left this is because @lorenzo left it in as well. I guess there's still a use for this. https://github.com/cakephp/cakephp/blob/7c3647c2776431b2fce8e0fa3a0e184c0a13a0ec/src/Database/Type.php#L57

markstory replied:
I suspect that is a mistake, as the Type factory will prefer a concrete type over a basic one.

markstory commented:
Having a totally separate set of types for the database seems to add a good amount of complexity for very little benefit. I know that having a Driver dependency in the Datasource package is not something we can have, but I think the Driver dependency was introduced solely to support binary encoding in SqlServer (at least that is the only place I can find it being used). If we can figure out an alternative way to solve that issue, I think there could be simpler solution to be found.

Marlinc commented:
So @markstory and I have been talking on IRC and we came up with the idea to create a SQL server specific ```BinaryType``` that gets loaded by the ```Sqlserver``` driver. This made it possible to get rid of a lot of complexity.

I've addressed most of the concerns you were having.

Marlinc commented on src/Database/Driver/Sqlserver.php:
```diff
@@ -70,6 +72,10 @@ public function connect()
         $dsn = "sqlsrv:Server={$config['host']};Database={$config['database']};MultipleActiveResultSets=false";
         $this->_connect($dsn, $config);
 
+        if ((empty($config['typeOverride'])) || ($config['typeOverride'] === true)) {
```
This way its still possible to disable the custom type.

Marlinc committed:
Move the type class to the datasource component

Marlinc committed:
Move basic types to the datasource component

markstory commented on src/Database/Type.php:
```diff
@@ -9,292 +9,46 @@
- * going from PHP into database.
- */
-class Type
+class Type extends \Cake\Datasource\Type implements TypeInterface
```
Missing a class doc block.

markstory commented on src/Database/Type/SqlserverBinaryType.php:
```diff
@@ -0,0 +1,62 @@
+ *
+ * Use to convert binary data between PHP and the database types.
+ */
+class SqlserverBinaryType extends \Cake\Datasource\Type\BinaryType implements TypeInterface
```
We generally import classes that are being extended.

markstory replied:
The SQLServer specific type is a much nicer approach - nice work.

markstory commented on src/Database/TypeInterface.php:
```diff
@@ -0,0 +1,28 @@
+ */
+namespace Cake\Database;
+
+interface TypeInterface
```
Need a class doc.

markstory commented on src/Database/TypeInterface.php:
```diff
@@ -0,0 +1,28 @@
+     *
+     * @return mixed
+     */
+    public function toStatement($value);
```
What about the other methods (toPHP, toDatasource) shouldn't they be in the interface as well?

markstory commented on src/Database/TypeTrait.php:
```diff
@@ -0,0 +1,37 @@
+
+/**
+ * Encapsulates all conversion functions for values coming from database into PHP and
+ * going from PHP into database.
```
Doesn't seem to match what the class is doing now.

markstory commented on src/Datasource/Type.php:
```diff
@@ -0,0 +1,275 @@
+     * @param mixed $value value to be converted to database equivalent
+     * @return mixed
+     */
+    public function toDatasource($value)
```
We need to keep the old toDatabase() method around as well or userland extensions will break.

markstory commented on src/Datasource/Type/BinaryType.php:
```diff
@@ -67,16 +59,4 @@ public function toPHP($value, Driver $driver)
-     * @param Driver $driver The driver.
-     * @return int
-     */
-    public function toStatement($value, Driver $driver)
```
All of the signature changes are not going to be very backwards compatible.

markstory commented:
In order to keep backwards compatibility, all of the moved classes will need aliases defined.

Marlinc commented:
@markstory I'll do that today.

lorenzo commented:
The actual reason for introducing `$driver` in the API was not because of SQLServer. Originally I thought I would need it for deciding on the date format depending on the driver. Later on, I thought it would be a good idea for custom types to provide different implementations depending on the driver

markstory commented:
Closing as its been a while and this has quite a few conflicts issues. If the conflicts are resolved we can take another look at this.
